command/debug/debug.go

```diff
@@ -0,0 +1,613 @@
+		if err != nil {
+			return version, fmt.Errorf("could not create output directory: %s", err)
+		}
+	} else {
+		return version, fmt.Errorf("output directory already exists: %s", c.output)
+	}
+
+	return version, nil
+}
+
+// captureStatic captures static target information and writes it
+// to the output path
+func (c *cmd) captureStatic() error {
+	// Collect errors via multierror as we want to gracefully
+	// fail if an API is inacessible
+	var errors error
+
+	// Collect the named outputs here
+	outputs := make(map[string]interface{}, 0)
+
+	// Capture host information
+	if c.configuredTarget("host") {
+		host, err := c.client.Agent().Host()
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+		outputs["host"] = host
+	}
+
+	// Capture agent information
+	if c.configuredTarget("agent") {
+		agent, err := c.client.Agent().Self()
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+		outputs["agent"] = agent
+	}
+
+	// Capture cluster members information, including WAN
+	if c.configuredTarget("cluster") {
+		members, err := c.client.Agent().Members(true)
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+		outputs["members"] = members
+	}
+
+	// Write all outputs to disk as JSON
+	for output, v := range outputs {
+		marshaled, err := json.MarshalIndent(v, "", "\t")
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+
+		err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", c.output, output), marshaled, 0644)
+		if err != nil {
+			errors = multierror.Append(errors, err)
+		}
+	}
+
+	return errors
+}
+
+// captureDynamic blocks for the duration of the command
+// specified by the duration flag, capturing the dynamic
+// targets at the interval specified
+func (c *cmd) captureDynamic() error {
+	successChan := make(chan int64)
+	errCh := make(chan error)
+	durationChn := time.After(c.duration)
+	intervalCount := 0
+
+	c.UI.Output(fmt.Sprintf("Beginning capture interval %s (%d)", time.Now().Local().String(), intervalCount))
+
+	// We'll wait for all of the targets configured to be
+	// captured before continuing
+	var wg sync.WaitGroup
+
+	capture := func() {
+		timestamp := time.Now().Local().Unix()
+
+		// Make the directory that will store all captured data
+		// for this interval
+		timestampDir := fmt.Sprintf("%s/%d", c.output, timestamp)
+		err := os.MkdirAll(timestampDir, 0755)
+		if err != nil {
+			errCh <- err
+		}
+
+		// Capture metrics
+		if c.configuredTarget("metrics") {
+			wg.Add(1)
+
+			go func() {
+				metrics, err := c.client.Agent().Metrics()
+				if err != nil {
+					errCh <- err
+				}
+
+				marshaled, err := json.MarshalIndent(metrics, "", "\t")
+				if err != nil {
+					errCh <- err
+				}
+
+				err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", timestampDir, "metrics"), marshaled, 0755)
+				if err != nil {
+					errCh <- err
+				}
+
+				// Sleep as other dynamic targets wait collect for the whole interv	al
+				time.Sleep(c.interval)
```

Comment:
Why does it matter that this goroutine stays alive the whole time if all it does is call Done() after that sleep?

Reply:
Great catch, thanks! That was before using wait groups.

Reply:
Actually I prematurely removed this. But as it was I think it was a bit of a gotcha. See the reasoning for keeping it (but moving it) in this commit: https://github.com/hashicorp/consul/pull/4754/commits/e4623c57a3e54d1a88d72281e43458d0c9d644f3

Reply:
Actually, scratch that altogether. In the case of only metrics being captured this is necessary -- the solution in the commit above (now removed) was also not accounting for other collections happening.

I think this needs to stay as-is.

Reply:
Cool, could you maybe update the comment to explain why? I wasn't really clear what the distinction between interval and duration were or that this was in a func that ends up being called in a loop.

Even having figured that out (makes sense) it's still somewhat subtle that we have to do this here. I'm not sure I grok why the commit above was wrong. Not a big deal but a comment might help  😄